### PR TITLE
Loss

### PR DESCRIPTION
**Pull request by clementd08**
Hello all,
This pull request introduce a new part for material in GUI, Losses.

297082554-b181b61d-ae13-4a08-a3e2-23945010bd10
297082617-5d2469dd-8090-4200-81e6-032db4a756c1
Now, you can modified matrix of loss in GUI.

Best regards,
Clément

### CODE REVIEW

File: Tests/GUI/Dialog/DMachineSetup/test_SSimu.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        self.widget = SSimu(machine=self.machine, material_dict=None, is_stator=None)
+
+    @classmethod
+    def teardown_class(cls):
+        """Exit the app after the test"""
+        cls.app.quit()
+
+    def test_set_N0(self):
+        """Check that the Widget allow to update N0"""
+        self.widget.lf_N0.clear()  # Clear the field before writing
```
Comment (BonneelP) [inline Tests/GUI/Dialog/DMachineSetup/test_SSimu.py:45]: Check init state

File: Tests/GUI/Dialog/DMachineSetup/test_SSimu.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        self.widget.g_losses_model.setChecked(True)
+
+        self.widget.lf_Trot.clear()  # Clear the field before writing
+        QTest.keyClicks(self.widget.lf_Trot, "130")
+        self.widget.lf_Trot.editingFinished.emit()  # To trigger the slot
+
+        assert self.widget.lf_Trot.value() == 130
+        assert self.widget.simu.loss.Trot == 130
+
+
```
Comment (BonneelP) [inline Tests/GUI/Dialog/DMachineSetup/test_SSimu.py:153]: Add test on run => Check folder creation + name of all files + number of files

File: Tests/GUI/Dialog/DMatLib/test_DMatSetup.py [hunk 3]
```diff
@@ -609,11 +625,156 @@ def test_BH_setup(self):
...
+        assert self.widget.w_setup.mat.mag.LossData.value[0, 2] == 20
+        assert self.widget.w_setup.mat.mag.LossData.value[1, 3] == 5
+        assert isfile(csv_path_Loss)
+        # Load csv
+        with mock.patch(
+            "PySide2.QtWidgets.QFileDialog.getOpenFileName",
+            return_value=(csv_path_Loss, "CSV (*.csv)"),
+        ):
+            mat.b_import.clicked.emit()
+            mat.emit_save()
```
Comment (BonneelP) [inline Tests/GUI/Dialog/DMatLib/test_DMatSetup.py:661]: emit save ?

File: Tests/GUI/Dialog/DMatLib/test_DMatSetup.py [hunk 4]
```diff
@@ -609,11 +625,156 @@ def test_BH_setup(self):
...
+        assert mat.w_tab.cellWidget(0, 2).value() == 0
+        assert mat.w_tab.cellWidget(1, 2).value() == 0.01
+        assert mat.w_tab.cellWidget(2, 2).value() == 0.02
+        assert mat.w_tab.cellWidget(3, 2).value() == 0.03
+        assert mat.w_tab.cellWidget(49, 2).value() == 0.49
+
+        assert self.widget.w_setup.mat.mag.LossData.value[1, 3] == 3
+        assert self.widget.w_setup.mat.mag.LossData.value[0, 3] == 50
+        assert self.widget.w_setup.mat.mag.LossData.value[2, 3] == 0.03
+
```
Comment (BonneelP) [inline Tests/GUI/Dialog/DMatLib/test_DMatSetup.py:770]: add click on plot

Comment (BonneelP) [reply Tests/GUI/Dialog/DMatLib/test_DMatSetup.py:770]: fig, ax to catch the plot
check that it is set, check figure/ax title + number lines/legend

File: pyleecan/GUI/Dialog/DMachineSetup/SSimu/SSimu.py [hunk 5]
```diff
@@ -134,6 +135,20 @@ def __init__(self, machine, material_dict, is_stator):
         if "MAIN" in config_dict and "RESULT_DIR" in config_dict["MAIN"]:
             self.w_path_result.set_path_txt(config_dict["MAIN"]["RESULT_DIR"])
 
+        # setup Losses Model
+        if isinstance(self.machine, MachineSIPMSM) or isinstance(
+            self.machine, MachineIPMSM
+        ):
+            self.g_losses_model.show()
+            if isinstance(self.simu.loss, LossFEA):
```
Comment (BonneelP) [inline pyleecan/GUI/Dialog/DMachineSetup/SSimu/SSimu.py:143]: no need for this if

File: pyleecan/GUI/Dialog/DMachineSetup/SSimu/SSimu.py [hunk 6]
```diff
@@ -287,6 +306,16 @@ def run(self):
...
+        # Losses
+        try:
+            out.loss.plot_losses(
+                is_show_fig=False, save_path=(join(self.simu.path_result, "Losses.png"))
+            )
+        except Exception as e:
+            err_msg = "Error while plotting Losses: " + str(e)
+            self.simu.get_logger().error(err_msg)
+
         # Done
```
[the remark this answers is not in the document]
Comment (BonneelP) [reply pyleecan/GUI/Dialog/DMachineSetup/SSimu/SSimu.py:319]: Check if meshsolution in output.h5
Find a way to add meshsolution to .mat => Adrien (export mag and losses meshsolution)

[the remark this answers is not in the document]
Comment (BonneelP) [reply pyleecan/GUI/Dialog/DMachineSetup/SSimu/SSimu.py:319]: add checkbox in magnetic "Export mesh and solution" (automatically checked+disabled if loss is checked) => is_get_meshsolution

File: pyleecan/GUI/Dialog/DMatLib/DMatSetup/DMatSetup.py [hunk 7]
```diff
@@ -292,6 +295,66 @@ def set_material(self, material, is_save_needed=False):
...
+            except Exception as e:
+                logger = getLogger(GUI_LOG_NAME)
+                logger.error(
+                    "Unable to import losses from Excel for "
+                    + str(self.mat.name)
+                    + ":\n"
+                    + str(e),
+                )
+                self.mat.mag.LossData = ImportMatrixVal(array([[0, 0, 0]]))
+            self.tab_values_losses.data = transpose(self.mat.mag.LossData.get_data())
```
Comment (BonneelP) [inline pyleecan/GUI/Dialog/DMatLib/DMatSetup/DMatSetup.py:345]: transpose add if

File: pyleecan/GUI/Dialog/DMatLib/DMatSetup/DMatSetup.py [hunk 8]
```diff
@@ -292,6 +295,66 @@ def set_material(self, material, is_save_needed=False):
...
+        self.tab_values_losses.si_row.show()
+
+        self.tab_values_losses.b_plot.show()
+        self.tab_values_losses.button_plot_title = "Loss(B)"
+
+        self.tab_values_losses.b_close.hide()
+        self.tab_values_losses.b_import.setHidden(False)
+        self.tab_values_losses.b_export.setHidden(False)
+
+        if isinstance(self.mat.mag.LossData, ImportMatrixXls):
```
Comment (BonneelP) [inline pyleecan/GUI/Dialog/DMatLib/DMatSetup/DMatSetup.py:330]: add comment why transpose

File: pyleecan/GUI/Tools/WTableData/DTableData.py [hunk 9]
```diff
@@ -159,7 +159,7 @@ def update_tab(self):
                 # Connect the slot
                 self.w_tab.cellWidget(ii, jj).editingFinished.connect(self.emit_save)
 
-        if ncol == 2:
+        if ncol == 2 or self.button_plot_title == "Loss(B)":
```
Comment (BonneelP) [inline pyleecan/GUI/Tools/WTableData/DTableData.py:162]: Nope

File: pyleecan/GUI/Tools/WTableData/DTableData.py [hunk 10]
```diff
@@ -352,6 +352,40 @@ def s_plot(self):
             axes.set_xlabel(self.col_header[0])
             axes.set_ylabel(self.col_header[1])
             fig.show()
+
+        elif (
```
Comment (BonneelP) [inline pyleecan/GUI/Tools/WTableData/DTableData.py:356]: Hide DTableData plot button, add plot button in DMatSetup, plot method to move in DMatSetup

File: pyleecan/Methods/Output/OutLoss/plot_losses.py [hunk 11]
```diff
@@ -11,6 +11,8 @@ def plot_losses(self, N0_array=None):
         An OutLoss object
     N0_array : ndarray
         speed array to use as X axis [rpm]
+    save_path : str
+        full path including folder, name and extension of the file to save if save_path is not None
     """
```
[the remark this answers is not in the document]
Comment (BonneelP) [reply pyleecan/Methods/Output/OutLoss/plot_losses.py:16]: is_show_fig default is True

File: Tests/GUI/Dialog/DMatLib/test_DMatSetup.py (hunk 4 above)
Comment (clementd08) [reply Tests/GUI/Dialog/DMatLib/test_DMatSetup.py:770]: self.fig is None
plot
self.fig is not None 
Check axes content
self.fig.close()

File: pyleecan/GUI/Dialog/DMatLib/DMatSetup/DMatSetup.py [hunk 12]
```diff
@@ -885,3 +954,88 @@ def change_type_material(self):
...
+            A DMatSetup object
+
+        Returns
+        -------
+        None
+        """
+        try:
+            data = self.tab_values_losses.data
+        except Exception as e:
+            QMessageBox.critical(self, self.tr("Error"), str(e))
```
[the remark this answers is not in the document]
Comment (BonneelP) [reply pyleecan/GUI/Dialog/DMatLib/DMatSetup/DMatSetup.py:999]: Functions.GUI.log_error

File: pyleecan/GUI/Dialog/DMatLib/DMatSetup/DMatSetup.py [hunk 13]
```diff
@@ -885,3 +954,88 @@ def change_type_material(self):
...
+                if a != data[temp, 0]:
+                    one_curve = False
+                    self.ax.plot(
+                        data[end:temp, 1], data[end:temp, 2], label=f" {a} [Hz]"
+                    )
+                    end = temp
+                    a = data[temp, 0]
+
+            if one_curve:
+                self.ax.plot(data[:, 1], data[:, 2], label=f" {data[0,0]} [Hz]")
```
Comment (BonneelP) [inline pyleecan/GUI/Dialog/DMatLib/DMatSetup/DMatSetup.py:1020]: Format freq to remove ".0"

File: pyleecan/Methods/Mesh/MeshSolution/save_mesh.py [hunk 14]
```diff
@@ -0,0 +1,62 @@
+# -*- coding: utf-8 -*-
+from scipy.io import savemat
+
+
+def save_mesh(self, save_path):
```
Comment (BonneelP) [inline pyleecan/Methods/Mesh/MeshSolution/save_mesh.py:5]: rename to "export_to_mat"

File: pyleecan/Methods/Mesh/MeshSolution/save_mesh.py [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+
+    Parameters
+    ----------
+    self : MeshSolution
+        an MeshSolution object
+    save_path : str
+        full path including folder, name and extension of the file to save if save_path is not None
+
+    """
+
```
Comment (BonneelP) [inline pyleecan/Methods/Mesh/MeshSolution/save_mesh.py:16]: check isdir => raise Exception
Check path ends with ".mat" if no extension add it, else raise exception

File: pyleecan/GUI/Dialog/DMachineSetup/SSimu/SSimu.py [hunk 16]
```diff
@@ -287,14 +303,81 @@ def run(self):
...
-            self,
-            self.tr("Simlation finished"),
+        self.test_err_msg = (
             "Simulation "
             + self.simu.name
             + " is finished.\nResults available at "
-            + self.simu.path_result,
+            + self.simu.path_result
+        )
+        QMessageBox().information(
```
Comment (BonneelP) [inline pyleecan/GUI/Dialog/DMachineSetup/SSimu/SSimu.py:377]: log_error

File: pyleecan/GUI/Dialog/DMatLib/DMatSetup/DMatSetup.py [hunk 17]
```diff
@@ -16,6 +16,10 @@
 from .....Functions.path_tools import rel_file_path
 from .....loggers import GUI_LOG_NAME
 
+import matplotlib.pyplot as plt
+from .....Functions.Plot.set_plot_gui_icon import set_plot_gui_icon
+from decimal import *
```
Comment (BonneelP) [inline pyleecan/GUI/Dialog/DMatLib/DMatSetup/DMatSetup.py:21]: to remove
%4e ? => cf FloatEdit

File: pyleecan/GUI/Dialog/DMatLib/DMatSetup/DMatSetup.py [hunk 18]
```diff
@@ -885,3 +955,104 @@ def change_type_material(self):
...
+            A DMatSetup object
+
+        Returns
+        -------
+        None
+        """
+        try:
+            data = self.tab_values_losses.data
+        except Exception as e:
+            error_message = f"Error while plotting losses, {e}"
```
Comment (BonneelP) [inline pyleecan/GUI/Dialog/DMatLib/DMatSetup/DMatSetup.py:1000]: self.test_err_msg

File: pyleecan/Methods/Mesh/MeshMat/save_vtk.py [hunk 19]
```diff
@@ -0,0 +1,42 @@
+import os
+from ....Functions.GUI.log_error import log_error
+
+
+def save_vtk(self, save_path):
+    """Save mesh to a .vtk file
+
+    Parameters
+    ----------
+    self : Mesh
```
Comment (BonneelP) [inline pyleecan/Methods/Mesh/MeshMat/save_vtk.py:10]: We already know how to save vtk in Manatee => Ask Hélène how we save the mesh in Manatee
add save_vtk method to MeshVTK (if needed)

File: pyleecan/Methods/Mesh/MeshSolution/export_to_mat.py [hunk 20]
```diff
@@ -0,0 +1,89 @@
+# -*- coding: utf-8 -*-
+from scipy.io import savemat
+import os
+from ....Functions.GUI.log_error import log_error
+
+
+def export_to_mat(self, save_path):
+    """Save mesh to a .mat file
```
Comment (BonneelP) [inline pyleecan/Methods/Mesh/MeshSolution/export_to_mat.py:8]: Save mesh and all solution

File: pyleecan/Methods/Mesh/MeshSolution/export_to_mat.py [hunk 21]
```diff
@@ -0,0 +1,89 @@
...
+                self,
+                err_msg,
+                self.mesh.get_logger(),
+                is_popup=False,
+                is_warning=False,
+            )
+
+    # Check if save_path exist
+    save_path.replace("\\", "/")
+    save_path_L = save_path.split("/")
```
Comment (BonneelP) [inline pyleecan/Methods/Mesh/MeshSolution/export_to_mat.py:35]: os or os.path => basedir, basename, splitpath (to get folder of file path)

Comment (BonneelP) [reply pyleecan/Methods/Mesh/MeshSolution/export_to_mat.py:35]: os.path.join

Comment (BonneelP) [reply pyleecan/Methods/Mesh/MeshSolution/export_to_mat.py:35]: isdir, isfile

File: pyleecan/Methods/Mesh/MeshSolution/export_to_mat.py [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+        raise KeyError("Error save path doesn't exist")
+
+    # dict to save
+    mesh_dict = dict()
+    mesh_dict["mesh"] = {}
+    mesh_dict["mesh"]["points_coordinates"] = self.mesh.get_mesh_pv().points
+    mesh_dict["mesh"]["cells_connectivity"] = self.mesh.get_mesh_pv().cells
+
+    mesh_dict["solution"] = {}
+    for value in self.solution_dict:
```
Comment (BonneelP) [inline pyleecan/Methods/Mesh/MeshSolution/export_to_mat.py:51]: for key in ...

File: pyleecan/Methods/Output/Output/plot/Magnetic/plot_B_mesh.py [hunk 23]
```diff
@@ -64,6 +65,8 @@ def plot_B_mesh(
     is_show_fig : bool
         To call show at the end of the method
     is_animated : True to animate magnetic flux density
+    is_2D : bool
+        True plot in 2 dimensions, False stayed in 3 Dimensions
```
Comment (BonneelP) [inline pyleecan/Methods/Output/Output/plot/Magnetic/plot_B_mesh.py:69]: True plot in 2 dimensions (view on XY)

File: pyleecan/GUI/Dialog/DMachineSetup/SSimu/SSimu.py [hunk 24]
```diff
@@ -221,7 +235,13 @@ def run(self):
             )
         except Exception as e:
             err_msg = "Error while plotting torque as fct of time: " + str(e)
```
Comment (BonneelP) [inline pyleecan/GUI/Dialog/DMachineSetup/SSimu/SSimu.py:237]: self.test_err_msg

File: pyleecan/Methods/Mesh/MeshMat/save_vtk.py [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+                self.get_logger(),
+                is_popup=False,
+                is_warning=False,
+            )
+
+    # Check if save_path exist
+    save_path_2 = os.path.dirname(save_path)
+
+    if not os.path.isdir(save_path_2):
+        raise KeyError("Error save path doesn't exist")
```
Comment (BonneelP) [inline pyleecan/Methods/Mesh/MeshMat/save_vtk.py:35]: KeyError => Exception
Error while saving mesh to vtk, save path doesn't exist: + path

File: pyleecan/Methods/Mesh/MeshSolution/export_to_mat.py [hunk 26]
```diff
@@ -0,0 +1,84 @@
...
+                self.mesh.get_logger(),
+                is_popup=False,
+                is_warning=False,
+            )
+
+    # Check if save_path exist
+    save_path_2 = os.path.dirname(save_path)
+
+    if not os.path.isdir(save_path_2):
+        raise KeyError("Error save path doesn't exist")
```
Comment (BonneelP) [inline pyleecan/Methods/Mesh/MeshSolution/export_to_mat.py:37]: same as save_vtk

File: pyleecan/Methods/Mesh/MeshVTK/save_vtk.py [hunk 27]
```diff
@@ -0,0 +1,38 @@
...
+                self.get_logger(),
+                is_popup=False,
+                is_warning=False,
+            )
+
+    # Check if save_path exist
+    save_path_2 = os.path.dirname(save_path)
+
+    if not os.path.isdir(save_path_2):
+        raise KeyError("Error save path doesn't exist")
```
Comment (BonneelP) [inline pyleecan/Methods/Mesh/MeshVTK/save_vtk.py:36]: same

File: Tests/GUI/Dialog/DMachineSetup/NewMachine/test_new_machine_Zoe.py [hunk 28]
```diff
@@ -68,6 +68,7 @@ def setup_method(self):
         """Run at the begining of every test to setup the gui"""
         # MatLib widget
         material_dict = load_matlib(matlib_path=matlib_path)
+        self.test_config_dict = {"GUI": {}, "MAIN": {}, "PLOT": {}}
```
Comment (BonneelP) [inline Tests/GUI/Dialog/DMachineSetup/NewMachine/test_new_machine_Zoe.py:71]: to remove

**Pull request merged**